File: julia/2022/day_05.jl
import Base.Iterators as It

const num_stacks = 9

function parse_stacks(str::AbstractString)::Vector{Vector{Char}}
    lines = split(str, '\n')
    stacks = [[] for _ in 1:num_stacks]
    # Loop starting from the bottom, skipping the index line
    for line in It.drop(It.reverse(lines), 1)
        # Loop over each crate column
        for (i, m) in enumerate(eachmatch(r"(?:\[(\w)\]|   ) ?", line))
            if !isnothing(m[1])
                push!(stacks[i], only(m[1]))
            end
        end
    end
    stacks
end

function parse_step(str::AbstractString)::Tuple{UInt, UInt, UInt}
    m = match(r"move (\d+) from (\d) to (\d)", str)
    tuple((parse(UInt, s) for s in m)...)
end

function do_steps(
    stacks::Vector{Vector{Char}},
    steps::AbstractVector{Tuple{UInt, UInt, UInt}};
    advanced::Bool
)::Vector{Vector{Char}}
    stacks = deepcopy(stacks)
    for (quantity, origin_i, dest_i) in steps
        origin, dest, = stacks[origin_i], stacks[dest_i]
        for _ in 1:quantity
            push!(dest, pop!(origin))
        end
        if advanced
            reverse!(dest, length(dest) - quantity + 1)
        end
    end
    stacks
end

function stack_tops(stacks::Vector{Vector{Char}})::String
    string(last.(stacks)...)
end

function main()
    text = read("input/2022/input_05.txt", String)
    state_str, steps_str = split(text, "\n\n")
    stacks = parse_stacks(state_str)
    steps = eachsplit(steps_str, '\n', keepempty=false) .|> parse_step
    for (p, advanced) in enumerate([false, true])
        solution = do_steps(stacks, steps, advanced=advanced) |> stack_tops
        println("Part $p: $solution")
    end
end

main()

File: rust/2022/src/bin/day_05.rs
use aoc::Part;

const NUM_STACKS: usize = 9;
const CRATE_WIDTH: usize = 3;

struct Step {
    amount: usize,
    source: usize,
    dest: usize,
}

fn parse_stacks(input: &str) -> [Vec<char>; NUM_STACKS] {
    let rows = input.lines().rev().skip(1);
    let mut stacks: [Vec<_>; NUM_STACKS] = Default::default();
    for row in rows {
        for (chunk, stack) in row.as_bytes().chunks(CRATE_WIDTH + 1).zip(&mut stacks) {
            let val = char::from(chunk[1]);
            if val.is_alphabetic() {
                stack.push(val);
            }
        }
    }
    stacks
}

fn parse_step(line: &str) -> Step {
    let mut nums = line.split_whitespace().filter_map(|s| s.parse().ok());
    Step {
        amount: nums.next().unwrap(),
        source: nums.next().unwrap() - 1,
        dest: nums.next().unwrap() - 1,
    }
}

fn do_steps(stacks: &mut [Vec<char>], steps: &[Step], part: Part) {
    for step in steps {
        let source = &mut stacks[step.source];
        let selected = source.split_off(source.len() - step.amount);
        let dest = &mut stacks[step.dest];
        let original_len = dest.len();
        dest.extend_from_slice(&selected);
        if part == Part::One {
            dest[original_len..].reverse();
        }
    }
}

fn main() {
    let input = aoc::read_input("2022/input_05.txt");
    let (stack_input, steps_input) = input.split_once("\n\n").unwrap();
    let steps: Vec<_> = steps_input.lines().map(parse_step).collect();
    for part in [Part::One, Part::Two] {
        let mut stacks = parse_stacks(stack_input);
        do_steps(&mut stacks, &steps, part);
        print!("{part}: ");
        for stack in stacks {
            print!("{}", stack.last().unwrap());
        }
        println!();
    }
}
